### PR TITLE
pallet-lockdrop: a module of lock dropping mechanism for bootstrapping other blockchains

### PR DESCRIPTION
**Pull request by sorpaas**
As much as we want to concentrate conducting experiments on the Kulupu blockchain, it's not practical to have one-blockchain-rule-it-all. The problem of bootstrapping a new blockchain inevitably comes.

For bootstrapping a PoW blockchain, it is easy -- we can just have no pre-mine, and the network will handle everything else by its own. However, things are slightly more complicated for a PoS blockchain, as pre-mine (or let's call it genesis allocation) cannot be avoided.

This PR proposes a new runtime module, called `pallet-lockdrop`, that provides a generic mechanism for people to use Kulupu to bootstrap other blockchains, with Kulupu community's permission.

The process works as follows:

* Once this module is enabled, anyone can propose, to Kulupu's democracy module, referendums to initiate a new lockdrop campaign. The campaign would specify some basic information.
* Kulupu community can vote to approve or reject the campaign. If rejected, the lockdrop will not happen on Kulupu blockchain.
* If approved, all KLP token holders can call `lockdrop::lock` to voluntarily lock their coins, for a specific amount of time. The lock period will then translate to genesis allocations on the other blockchain.
* When locked, the coin is not transferrable, but the KLP holder can still use it to participate in Kulupu governance.
* Once the lock period expires, one can call `lockdrop::unlock` to unlock the coins.

The specific instance right now we have, for a concrete lockdrop process, is [Neatcoin's bootstrapping process](https://neatcoin.org/docs/bootstrap). However, note that this PR only proposes the generic mechanism. Even if this PR landed on-chain, you still have a second chance to vote on whether the particular instance (Neatcoin's bootstrapping) will happen or not.

### CODE REVIEW

File: frame/lockdrop/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+                }
+            }
+        }
+
+        #[weight = 0]
+        fn lock(origin, amount: BalanceOf<T>, identifier: CampaignIdentifier, lock_end_block: T::BlockNumber) {
+            let account_id = ensure_signed(origin)?;
+            let campaign_info = Campaigns::<T>::get(&identifier).ok_or(Error::<T>::CampaignNotExists)?;
+            
+            ensure!(lock_end_block > campaign_info.min_lock_end_block, Error::<T>::InvalidLockEndBlock);
```
Comment (shawntabrizi) [inline frame/lockdrop/src/lib.rs:129]: should also check that the current block is before the end block

File: frame/lockdrop/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct CampaignInfo<T: Config> {
+    end_block: T::BlockNumber,
+    min_lock_end_block: T::BlockNumber,
+}
+
+#[derive(Encode, Decode, Clone, Eq, PartialEq, RuntimeDebug)]
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+pub struct LockInfo<T: Config> {
+    end_block: T::BlockNumber,
```
Comment (shawntabrizi) [inline frame/lockdrop/src/lib.rs:46]: i think to make this pallet more helpful, also holding the amount locked is a good idea.

**Comment by shawntabrizi** [comment]
Creating a using a child-trie here makes sense so that downstream chains can use the root hash and merkle proofs for people to claim their tokens on the appropriate chains. This is the same thing crowdloans use in Polkadot.

File: frame/lockdrop/src/lib.rs (hunk 1 above)
Comment (sorpaas) [reply frame/lockdrop/src/lib.rs:136]: Ooops!

**Comment by sgaragagghu** [comment]
Maybe I lost something but doesn't make sense to have the start lock block in the child trie ? 🤔

**Comment by sorpaas** [comment]
@sgaragagghu The trie is only there for some chains who might use the proof/claim process. We don't need to put everything in the trie. If needed, the start lock block can always be retrieved from the root trie.

**Comment by JackZhou2012** [comment]
锁定空投模块是个非常好的模块！

但是要想使用这个模块还需要社区投票这一点有点繁琐，会阻碍人们使用。同时让社区投票赞成或反对，这样就给社区带来了鉴别项目好坏的任务压力。如果投票赞成，一旦项目出现骗局，社区会不可避免的会承担一部分责任，会给社区带来负面的影响。

所以，我建议去掉投票赞成反对这一步，这样我们既没有鉴别项目的任务，也不需要去承担任何责任。

任何人都可以在Kulupu上发起锁定空投的项目，无需社区同意，只要拥有KLP都可以来锁定空投。把Kulupu建设得更加开放，就像以太坊那样，让更多的人可以自由使用。

**Comment by JackZhou2012** [comment]
> As much as we want to concentrate conducting experiments on the Kulupu blockchain, it's not practical to have one-blockchain-rule-it-all. The problem of bootstrapping a new blockchain inevitably comes.
> 
> For bootstrapping a PoW blockchain, it is easy -- we can just have no pre-mine, and the network will handle everything else by its own. However, things are slightly more complicated for a PoS blockchain, as pre-mine (or let's call it genesis allocation) cannot be avoided.
> 
> This PR proposes a new runtime module, called `pallet-lockdrop`, that provides a generic mechanism for people to use Kulupu to bootstrap other blockchains, with Kulupu community's permission.
> 
> The process works as follows:
> 
> * Once this module is enabled, anyone can propose, to Kulupu's democracy module, referendums to initiate a new lockdrop campaign. The campaign would specify some basic information.
> * Kulupu community can vote to approve or reject the campaign. If rejected, the lockdrop will not happen on Kulupu blockchain.
> * If approved, all KLP token holders can call `lockdrop::lock` to voluntarily lock their coins, for a specific amount of time. The lock period will then translate to genesis allocations on the other blockchain.
> * When locked, the coin is not transferrable, but the KLP holder can still use it to participate in Kulupu governance.
> * Once the lock period expires, one can call `lockdrop::unlock` to unlock the coins.
> 
> The specific instance right now we have, for a concrete lockdrop process, is [Neatcoin's bootstrapping process](https://neatcoin.org/docs/bootstrap). However, note that this PR only proposes the generic mechanism. Even if this PR landed on-chain, you still have a second chance to vote on whether the particular instance (Neatcoin's bootstrapping) will happen or not.

锁定空投模块是个非常好的模块！

但是要想使用这个模块还需要社区投票这一点有点繁琐，会阻碍人们使用。同时让社区投票赞成或反对，这样就给社区带来了鉴别项目好坏的任务压力。如果投票赞成，一旦项目出现骗局，社区会不可避免的会承担一部分责任，会给社区带来负面的影响。

所以，我建议去掉投票赞成反对这一步，这样我们既没有鉴别项目的任务，也不需要去承担任何责任。

任何人都可以在Kulupu上发起锁定空投的项目，无需社区同意，只要拥有KLP都可以来锁定空投。把Kulupu建设得更加开放，就像以太坊那样，让更多的人可以自由使用。

**Pull request merged**